src/common/exponent/exponentHelper.ts

```diff
@@ -105,305 +74,199 @@ export class ExponentHelper {
-                const fileContents =
-                    `// This file is automatically generated by ${extensionName}@${extensionVersionNumber}
-// Please do not modify it manually. All changes will be lost.
-var React = require('${this.projectRootPath}/node_modules/react');
-var {Component} = React;
+        return this.detectEntry()
+            .then((entryPoint: string) => {
+                const content = this.generateFileContent(name, entryPoint);
+                return this.fs.writeFile(this.pathToFileInWorkspace(EXPONENT_INDEX), content);
+            });
 
-var ReactNative = require('${this.projectRootPath}/node_modules/react-native');
-var {AppRegistry} = ReactNative;
+    }
 
-var entryPoint = require('${this.projectRootPath}/${entryPointFile}');
+    private detectEntry(): Q.Promise<string> {
+        this.lazilyInitialize();
+        return Q.all([
+            this.fs.exists(this.pathToFileInWorkspace(DEFAULT_EXPONENT_INDEX)),
+            this.fs.exists(this.pathToFileInWorkspace(DEFAULT_IOS_INDEX)),
+            this.fs.exists(this.pathToFileInWorkspace(DEFAULT_ANDROID_INDEX)),
+        ])
+        .spread((expo: boolean, ios: boolean, android: boolean): string => {
+            return expo ? this.pathToFileInWorkspace(DEFAULT_EXPONENT_INDEX) :
+                ios ? this.pathToFileInWorkspace(DEFAULT_IOS_INDEX) :
+                this.pathToFileInWorkspace(DEFAULT_ANDROID_INDEX);
+        });
+    }
 
+    private generateFileContent(name: string, entryPoint: string): string {
+        return `// This file is automatically generated by VS Code
+// Please do not modify it manually. All changes will be lost.
+var React = require('${this.pathToFileInWorkspace("/node_modules/react")}');
+var { Component } = React;
+var ReactNative = require('${this.pathToFileInWorkspace("/node_modules/react-native")}');
+var { AppRegistry } = ReactNative;
+var entryPoint = require('${entryPoint}');
 AppRegistry.registerRunnable('main', function(appParameters) {
-    AppRegistry.runApplication('${componentName}', appParameters);
+    AppRegistry.runApplication('${name}', appParameters);
 });`;
-                return this.fileSystem.writeFile(this.dotvscodePath(EXPONENT_INDEX), fileContents);
-            });
     }
 
-    /**
-     * Create exp.json file in the workspace root if not present
-     */
-    private ensureExpJson(): Q.Promise<void> {
-        this.lazilyInitialize();
-        let defaultSettings = {
-            "sdkVersion": "",
-            "entryPoint": this.dotvscodePath(EXPONENT_INDEX),
-            "slug": "",
-            "name": "",
-        };
-        return this.readVscodeExponentSettingFile()
-            .then(exponentJson => {
-                const expJsonPath = this.pathToFileInWorkspace(EXP_JSON);
-                if (!this.fileSystem.existsSync(expJsonPath) || exponentJson.overwriteExpJson) {
+    private patchAppJson(isExpo: boolean = true): Q.Promise<void> {
+        return this.readAppJson()
```

Comment:
I can see at least one code path where the exception here will not be caught by caller code which likely would lead to extension crash - this is a bad UX and must be avoided if possible.

Have you tested this scenario?